### PR TITLE
Support Min P Sampler

### PR DESCRIPTION
**Pull request by esmeetu**
### Reference: 
https://github.com/oobabooga/text-generation-webui/pull/4449
https://github.com/ggerganov/llama.cpp/pull/3841

It seems sample better result than normal top_k/top_p method. And i tested min_p = 0.05 when inferencing code LLM, it gave me a better explanation.

### CODE REVIEW

File: vllm/model_executor/layers/sampler.py [hunk 1]
```diff
@@ -313,6 +322,28 @@ def _apply_top_p_top_k(
...
+    min_p = torch.tensor(min_ps, dtype=logits.dtype, device=logits.device)
+    probs = torch.softmax(logits, dim=-1)
+    top_probs, _ = probs.max(dim=-1, keepdim=True)
+    scaled_min_p = min_p.unsqueeze(dim=1) * top_probs
+    tokens_to_remove = probs < scaled_min_p
+
+    sorted_indices = torch.argsort(logits, descending=True, dim=-1)
+    sorted_indices_to_remove = torch.gather(tokens_to_remove,
+                                            dim=-1,
+                                            index=sorted_indices)
```
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:338]: Seems like the sort here is unnecessary? I believe we only need to mask out the positions where `probs < scaled_min_p`?

File: vllm/model_executor/layers/sampler.py [hunk 2]
```diff
@@ -313,6 +322,28 @@ def _apply_top_p_top_k(
     return logits
 
 
+def _apply_min_p(
+    logits: torch.Tensor,
+    min_ps: List[float],
+) -> torch.Tensor:
```
Comment (zhuohan123) [inline vllm/model_executor/layers/sampler.py:328]: Add a reference to the original implementation here?

**Comment by esmeetu** [comment]
@zhuohan123 Thanks for you reviews! In terms of first problem, i removed the `min_tokens_to_keep ` params which need sorting to take from original logits. Why i removed that params is because i want to keep this implement simpler, and align with top_p/top_k.

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thank you for your great contribution!
